api/src/services/opportunities_v0_1/search_opportunities.py

```diff
@@ -1,20 +1,117 @@
+import logging
 from typing import Sequence, Tuple
 
-from pydantic import BaseModel
-from sqlalchemy import asc, desc, select
-from sqlalchemy.orm import joinedload
+from pydantic import BaseModel, Field
+from sqlalchemy import Select, asc, desc, or_, select
+from sqlalchemy.orm import noload, selectinload
 
 import src.adapters.db as db
-from src.db.models.opportunity_models import Opportunity
+from src.db.models.opportunity_models import (
+    CurrentOpportunitySummary,
+    LinkOpportunitySummaryApplicantType,
+    LinkOpportunitySummaryFundingCategory,
+    LinkOpportunitySummaryFundingInstrument,
+    Opportunity,
+    OpportunitySummary,
+)
 from src.pagination.pagination_models import PaginationInfo, PaginationParams
 from src.pagination.paginator import Paginator
 
+logger = logging.getLogger(__name__)
+
+
+class SearchOpportunityFilters(BaseModel):
+    funding_instrument: dict | None = Field(default=None)
+    funding_category: dict | None = Field(default=None)
+    applicant_type: dict | None = Field(default=None)
+    opportunity_status: dict | None = Field(default=None)
+    agency: dict | None = Field(default=None)
 
-class SearchOpportunityParams(BaseModel):
-    # Filters will be added in a subsequent ticket
 
+class SearchOpportunityParams(BaseModel):
     pagination: PaginationParams
 
+    query: str | None = Field(default=None)
+    filters: SearchOpportunityFilters | None = Field(default=None)
+
+
+def _add_query_filters(
+    stmt: Select[tuple[Opportunity]], query: str | None
+) -> Select[tuple[Opportunity]]:
+    if query is None or len(query) == 0:
+        return stmt
+
+    # TODO - will implement this in https://github.com/HHS/simpler-grants-gov/issues/1455
+
+    return stmt
+
+
+def _add_filters(
+    stmt: Select[tuple[Opportunity]], filters: SearchOpportunityFilters | None
+) -> Select[tuple[Opportunity]]:
+    if filters is None:
+        return stmt
+
+    # We need to add joins so that the where clauses
+    # can query against the tables that are relevant for these filters
+    # Current + Opportunity Summary are always needed so just add them here
+    stmt = stmt.join(CurrentOpportunitySummary).join(
+        OpportunitySummary,
+        CurrentOpportunitySummary.opportunity_summary_id
+        == OpportunitySummary.opportunity_summary_id,
+    )
+
+    if filters.opportunity_status is not None:
+        one_of_opportunity_statuses = filters.opportunity_status.get("one_of")
+
+        if one_of_opportunity_statuses:
+            stmt = stmt.where(
+                CurrentOpportunitySummary.opportunity_status.in_(one_of_opportunity_statuses)
+            )
+
+    if filters.funding_instrument is not None:
+        stmt = stmt.join(LinkOpportunitySummaryFundingInstrument)
```

Comment:
Should this go inside the `if` below too? So that we only `join` if it's going to have a `where`.

Reply:
The idea is that if we have multiple filters for the field, it adds the join for it collectively, we just only have one for each right now.

I should however make it so the filter model makes it so you'll always have at least one of these set so that we don't ever have a join for no reason. I'll make a follow-up ticket to add that.